SimTKcommon/src/ParallelExecutor.cpp

```diff
@@ -66,36 +79,48 @@ ParallelExecutorImpl::~ParallelExecutorImpl() {
     pthread_cond_destroy(&waitCondition);
 }
 ParallelExecutorImpl* ParallelExecutorImpl::clone() const {
-    return new ParallelExecutorImpl(threads.size());
+    return new ParallelExecutorImpl(NUM_MAX_THREADS);
 }
 void ParallelExecutorImpl::execute(ParallelExecutor::Task& task, int times) {
-    if (times == 1 || threads.size() == 1) {
-        // Nothing is actually going to get done in parallel, so we might as well
-        // just execute the task directly and save the threading overhead.
-        
-        task.initialize();
-        for (int i = 0; i < times; ++i)
-            task.execute(i);
-        task.finish();
-        return;
-    }
-    
-    // Initialize fields to execute the new task.
-    
-    pthread_mutex_lock(&runLock);
-    currentTask = &task;
-    currentTaskCount = times;
-    waitingThreadCount = 0;
-    for (int i = 0; i < (int) threadInfo.size(); ++i)
-        threadInfo[i]->running = true;
-
-    // Wake up the worker threads and wait until they finish.
-    
-    pthread_cond_broadcast(&runCondition);
-    do {
-        pthread_cond_wait(&waitCondition, &runLock);
-    } while (waitingThreadCount < (int) threads.size());
-    pthread_mutex_unlock(&runLock);
+  if (min(times, NUM_MAX_THREADS) == 1) {
+      //(1) NON-PARALLEL CASE:
+      // Nothing is actually going to get done in parallel, so we might as well
+      // just execute the task directly and save the threading overhead.
+      task.initialize();
+      for (int i = 0; i < times; ++i)
+          task.execute(i);
+      task.finish();
+      return;
+  }else{
+      //(2) PARALLEL CASE:
+      // We launch the needed number of threads, accounting for the fact that
+      // we may have previously already launched a smaller number of threads
+      int numThreadsToLaunch = min(times, NUM_MAX_THREADS);
+      if(threads.size() < numThreadsToLaunch)
+      {
+        int prevThreadSize = threads.size();
+        threads.resize(numThreadsToLaunch);
```

Comment:
I can't explicitly write a test to make sure that the pointers are the same at each index after resizing since the Array_ of threads isn't exposed publicly, but I will look at it in the debugger. According to the Array_ documentation, resize should 

>>Change the size of this Array, preserving all the elements that will still 
fit, and default constructing any new elements that are added. This is not
allowed for non-owner arrays unless the requested size is the same as the 
current size.